The global state refactor (#422)
fission-codes/webnative: README.md

The Webnative SDK empowers developers to build fully distributed web applications without needing a complex back-end. The SDK provides:

- **User accounts** via the browser's [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) or by using a blockchain wallet as a [webnative plugin](https://github.com/fission-codes/webnative-walletauth).
- **Authorization** using [UCAN](https://ucan.xyz/).
- **Encrypted file storage** via the [Webnative File System](https://guide.fission.codes/developers/webnative/file-system-wnfs) backed by [IPLD](https://ipld.io/).
- **Key management** via websockets and a two-factor auth-like flow.

Webnative applications work offline and store data encrypted for the user by leveraging the power of the web platform. You can read more about Webnative in Fission's [Webnative Guide](https://guide.fission.codes/developers/webnative). There's also an API reference which can be found at [webnative.fission.app](https://webnative.fission.app)

# Getting started

```ts
// ESM
import * as wn from "webnative"

// Browser/UMD build
const wn = globalThis.webnative
```

## Creating a Program

> A Webnative program is an assembly that makes up a part of, or a whole distributed web application. It consists out of several components that can be customised. _We'll stick with the default components for now, which means we'll be using the Web Crypto API._

Author: Well, I would argue that "of components" is implied, one of definitions of assembly is:
> a unit consisting of components that have been fitted together

I agree with the other changes 👍

```ts
const program = await wn.program({
  // Can also be a string, used as an identifier for caches.
  // If you're developing multiple apps on the same localhost port,
  // make sure these differ.

>   id: { creator: "Nullsoft", name: "Winamp" }

Reviewer: Mentioned in the WAT PR the idea of using `namespace` in place of `tag` or `id`. what do you think?
```suggestion
  namespace: { creator: "Nullsoft", name: "Winamp" }
```

Author: Oh great idea! That's the word I was looking for 👌

Author: Updated to `namespace`

}).catch(error => {
  switch (error) {
    case webnative.ProgramError.InsecureContext:
      // Webnative requires HTTPS
      break;
    case webnative.ProgramError.UnsupportedBrowser:
      break;
  }

})
```

> That gives us a `Program` object, with this we can create a new user session or reuse the existing session. There's two distinctive ways to create a user session, either by using an authentication strategy or by using something we call "confidences". More on the latter in a bit, let's start with the default authentication strategy.

Reviewer 2: ```suggestion
`wn.program` returns a `Program` object, which can create a new user session or reuse an existing session. There are two ways to create a user session, either by using an authentication strategy or by requesting access from another app through the "confidences" system. Let's start with the default authentication strategy.
```

Possible rewording. Also added a suggestion for how we might hint at confidences and how they are different from selecting an authentication strategy.

```ts
let session

// Do we have an existing session?
if (program.session) {
  session = program.session

// If not, let's authenticate.
// (a) new user, register a new Fission account

> } else if (registerNewUser) {

Reviewer 2: How would a developer decide what to set `registerNewUser` to?

Author: It's decided by the user wether or not they have an account, eg. this screen in WAT:

<img width="393" alt="Screenshot 2022-11-18 at 11 23 59" src="https://user-images.githubusercontent.com/296665/202680102-4e4fc9ec-6eba-4a9f-9c9f-8ca8b6ad9e7c.png">

The wording could be better here yeah.
Maybe `userChoseToRegister`?

Reviewer 2: Yeah, that wording would be better! 👌

  const { success } = await program.auth.register({ username: "llama" })
  session = success ? program.auth.session() : null

// (b) existing user, link a new device
} else {
  // On device with existing session:
  const producer = program.auth.accountProducer(program.session.username)

  producer.on("challenge", challenge => {
    // Either show `challenge.pin` or have the user input a PIN and see if they're equal.
    if (userInput === challenge.pin) challenge.confirmPin()
    else challenge.rejectPin()
  })

  producer.on("link", ({ approved }) => {
    if (approved) console.log("Link device successfully")
  })

  // On device without session:

>   //     Somehow you'll need to get a hold of the username.

Reviewer: small typo. `ahold` is one word in this context :) 
```suggestion
  //     Somehow you'll need to get ahold of the username.
```

Author: Good to know, thanks 😄

  //     Few ideas: URL query param, QR code, manual input.
  const consumer = program.auth.accountConsumer(username)

  consumer.on("challenge", ({ pin }) => {
    showPinOnUI(pin)
  })

  consumer.on("link", ({ approved, username }) => {
    if (approved) {
      console.log(`Successfully authenticated as ${username}`)
      session = program.auth.session()
    }
  })
}
```

> Alternatively you can use "confidences", this system is used when you want partial access to a file system. At the moment of writing this only supported through the "Fission auth lobby", which is a Webnative app that uses the auth strategy shown above.

Reviewer 2: ```suggestion
Alternatively you can use the "confidences" system when you want partial access to a file system. At the moment of writing, confidences are only supported through the "Fission auth lobby", which is a Webnative app that uses the auth strategy shown above.
```

This Fission auth lobby flow works as follows:
1. You get redirected to the Fission lobby from your app.
2. Here you create an account like in the normal auth strategy flow shown above.
3. The lobby shows what your app wants to access in your file system.
4. You approve or deny these permissions and get redirected back to your app.
5. Your app collects the encrypted information (UCANs & file system secrets).

> 6. You can create a user session.

Reviewer 2: ```suggestion
6. Your app can create a user session.
```
To be consistent with the other ones.

```ts
// We define a `Permissions` object,
// this represents what permissions to ask the user.
const permissions = {
  // Ask permission to write to and read from the directory:
  // private/Apps/Nullsoft/Winamp
  app: { creator: "Nullsoft", name: "Winamp" }
}

// We need to pass this object to our program
const program = await webnative.program({
  permissions
})

// (a) Whenever you are ready to redirect to the lobby, call this:
program.confidences.request(permissions)

// (b) When you get redirected back and your program is ready,
// you will have access to your user session.
session = program.session
```

> Once you have your `Session`, you have access to the file system 🎉

Reviewer: Maybe use `your` in place of `the`, as it is your file system?
```suggestion
Once you have your `Session`, you have access to your file system 🎉
```

```ts
const fs = session.fs
```

__Notes:__

- You can use alternative authentication strategies, such as [webnative-walletauth](https://github.com/fission-codes/webnative-walletauth).

> - You can remove all traces of the user using `await session.destroy()`

Reviewer 2: Yeah, nice to have this exposed off the session. 🧨

> - You can load the file system separately, in case you're using web worker. This is done using the combination of `configuration.filesystem.loadImmediately = false` and `program.loadFileSystem()`

Reviewer 2: ```suggestion
- You can load the file system separately if you're using a web worker. This is done using the combination of `configuration.filesystem.loadImmediately = false` and `program.loadFileSystem()`
```

```

__Notes:__

- You can use alternative authentication strategies, such as [webnative-walletauth](https://github.com/fission-codes/webnative-walletauth).
- You can remove all traces of the user using `await session.destroy()`
- You can load the file system separately if you're using a web worker. This is done using the combination of `configuration.filesystem.loadImmediately = false` and `program.loadFileSystem()`

## Working with the file system

The Web Native File System (WNFS) is a file system built on top of [IPLD](https://ipld.io/). Each file system has a public side and a private side, much like your macOS, Windows, or Linux desktop file system. The public side is "live" and publicly accessible on the Internet. The private side is encrypted so that only the owner can see the contents. Read more about it [here](https://github.com/wnfs-wg).

```ts
const { Branch } = wn.path

// List the user's private files
await fs.ls(
  wn.path.directory(Branch.Private)
)

// Create a sub directory and add some content
const contentPath = wn.file(
  Branch.Private, "Sub Directory", "hello.txt"
)

await fs.write(
  contentPath,
  new TextEncoder().encode("👋") // Uint8Array
)

> // Announce the changes to your other devices

Author: I like it!
